File: apps/cli/internal/persistence/spend.go
package persistence

import (
	"database/sql"
	"fmt"
	"os"
	"path/filepath"
	"sync"
	"time"

	_ "github.com/ncruces/go-sqlite3/driver"
	_ "github.com/ncruces/go-sqlite3/embed"
)

const (
	spendDBFile           = "spend.db"
	spendSchemaVersion    = 1
)

// SpendDB handles global spend tracking across all repositories.
// Unlike SQLiteWriter which is per-repo, this tracks spend globally
// in ~/.detent/spend.db for accurate monthly budget enforcement.
type SpendDB struct {
	db   *sql.DB
	path string
	mu   sync.RWMutex
}

// spendDB is the singleton global spend database
var (
	globalSpendDB   *SpendDB
	globalSpendOnce sync.Once
	errGlobalSpend  error
)

// GetSpendDB returns the global spend database singleton.
// The database is created lazily and kept open for the lifetime of the process.
func GetSpendDB() (*SpendDB, error) {
	globalSpendOnce.Do(func() {
		globalSpendDB, errGlobalSpend = openSpendDB()
	})
	return globalSpendDB, errGlobalSpend
}

// openSpendDB opens (or creates) the global spend database.
func openSpendDB() (*SpendDB, error) {
	detentDir, err := GetDetentDir()
	if err != nil {
		return nil, err
	}

	// Ensure detent directory exists
	// #nosec G301 - 0700 is intentionally restrictive
	if mkdirErr := os.MkdirAll(detentDir, 0o700); mkdirErr != nil {
		return nil, fmt.Errorf("creating detent directory: %w", mkdirErr)
	}

	dbPath := filepath.Join(detentDir, spendDBFile)
	db, err := sql.Open("sqlite3", dbPath)
	if err != nil {
		return nil, fmt.Errorf("opening spend database: %w", err)
	}

	// Configure connection pool - single connection for SQLite
	db.SetMaxOpenConns(1)
	db.SetMaxIdleConns(1)
	db.SetConnMaxLifetime(0)

	// Apply performance pragmas
	pragmas := []string{
		"PRAGMA journal_mode=WAL",
		"PRAGMA synchronous=NORMAL",
		"PRAGMA busy_timeout=5000",
	}
	for _, pragma := range pragmas {
		if _, err := db.Exec(pragma); err != nil {
			_ = db.Close()
			return nil, fmt.Errorf("executing %s: %w", pragma, err)
		}
	}

	sdb := &SpendDB{db: db, path: dbPath}

	// Initialize schema
	if err := sdb.initSchema(); err != nil {
		_ = db.Close()
		return nil, fmt.Errorf("initializing spend schema: %w", err)
	}

	// Set secure permissions
	// #nosec G302 - intentionally restrictive permissions
	if err := os.Chmod(dbPath, 0o600); err != nil {
		_ = db.Close()
		return nil, fmt.Errorf("setting spend database permissions: %w", err)
	}

	return sdb, nil
}

// initSchema creates the spend_log table if needed.
func (s *SpendDB) initSchema() error {
	schema := `
	CREATE TABLE IF NOT EXISTS schema_version (
		version INTEGER PRIMARY KEY,
		applied_at INTEGER NOT NULL
	);

	CREATE TABLE IF NOT EXISTS spend_log (
		id INTEGER PRIMARY KEY AUTOINCREMENT,
		cost_usd REAL NOT NULL,
		repo_id TEXT,
		created_at INTEGER NOT NULL
	);

	CREATE INDEX IF NOT EXISTS idx_spend_log_created_at ON spend_log(created_at);
	`

	if _, err := s.db.Exec(schema); err != nil {
		return fmt.Errorf("creating spend schema: %w", err)
	}

	// Check/update schema version
	var version int
	if err := s.db.QueryRow("SELECT COALESCE(MAX(version), 0) FROM schema_version").Scan(&version); err != nil {
		return fmt.Errorf("querying schema version: %w", err)
	}

	if version < spendSchemaVersion {
		if _, err := s.db.Exec("INSERT INTO schema_version (version, applied_at) VALUES (?, ?)",
			spendSchemaVersion, time.Now().Unix()); err != nil {
			return fmt.Errorf("recording schema version: %w", err)
		}
	}

	return nil
}

// RecordSpend records a spend amount to the global spend log.
// repoID is optional and used for auditing purposes.
func (s *SpendDB) RecordSpend(costUSD float64, repoID string) error {
	if costUSD <= 0 {
		return nil
	}

	s.mu.Lock()
	defer s.mu.Unlock()

	query := `INSERT INTO spend_log (cost_usd, repo_id, created_at) VALUES (?, ?, ?)`
	_, err := s.db.Exec(query, costUSD, repoID, time.Now().Unix())
	if err != nil {
		return fmt.Errorf("recording spend: %w", err)
	}

	return nil
}

// GetMonthlySpend returns the total spend for the given month (format: "2006-01").
// If month is empty, uses the current month.
func (s *SpendDB) GetMonthlySpend(month string) (float64, error) {
	if month == "" {
		month = time.Now().Format("2006-01")
	}

	start, err := time.Parse("2006-01", month)
	if err != nil {
		return 0, fmt.Errorf("invalid month format: %w", err)
	}
	end := start.AddDate(0, 1, 0)

	s.mu.RLock()
	defer s.mu.RUnlock()

	query := `SELECT COALESCE(SUM(cost_usd), 0) FROM spend_log WHERE created_at >= ? AND created_at < ?`
	var total float64
	if err := s.db.QueryRow(query, start.Unix(), end.Unix()).Scan(&total); err != nil {
		return 0, fmt.Errorf("querying monthly spend: %w", err)
	}

	return total, nil
}

// Close closes the spend database connection.
func (s *SpendDB) Close() error {
	if s.db == nil {
		return nil
	}
	return s.db.Close()
}

File: packages/persistence/src/spend.ts
/**
 * Global spend tracking (SEPARATE DATABASE)
 * Ported from Go: apps/go-cli/internal/persistence/spend.go
 *
 * Unlike the per-repo database, the spend database is stored at ~/.detent/spend.db
 * and tracks spend globally across all repositories.
 */

import { chmodSync, existsSync, mkdirSync } from "node:fs";
import { homedir } from "node:os";
import { join } from "node:path";
import Database from "better-sqlite3";

// ============================================================================
// Constants
// ============================================================================

const SPEND_DB_FILE = "spend.db";
const DETENT_DIR_NAME = ".detent";
const WINDOWS_DRIVE_PATTERN = /^[A-Za-z]:\\/;

// ============================================================================
// Singleton Pattern
// ============================================================================

let globalSpendDb: Database.Database | null = null;

/**
 * Validates an override path from environment variable.
 * Returns null if path is invalid (contains traversal sequences or is not absolute).
 */
const validateOverridePath = (path: string): string | null => {
  // Reject paths with traversal sequences
  if (path.includes("..")) {
    return null;
  }

  // Reject non-absolute paths
  if (!(path.startsWith("/") || WINDOWS_DRIVE_PATTERN.test(path))) {
    return null;
  }

  return path;
};

/**
 * Gets the detent directory path (~/.detent)
 */
export const getDetentDir = (): string => {
  const override = process.env.DETENT_HOME;
  if (override) {
    const validated = validateOverridePath(override);
    if (validated) {
      return validated;
    }
    // Invalid override path - fall back to default
  }
  return join(homedir(), DETENT_DIR_NAME);
};

/**
 * Gets or creates the global spend database singleton
 */
export const getSpendDb = (): Database.Database => {
  if (globalSpendDb) {
    return globalSpendDb;
  }

  const detentDir = getDetentDir();

  // Ensure detent directory exists
  if (!existsSync(detentDir)) {
    mkdirSync(detentDir, { mode: 0o700, recursive: true });
  }

  const dbPath = join(detentDir, SPEND_DB_FILE);
  const db = new Database(dbPath);

  // Configure for optimal SQLite performance
  db.pragma("journal_mode = WAL");
  db.pragma("synchronous = NORMAL");
  db.pragma("busy_timeout = 5000");

  // Initialize schema
  initSpendSchema(db);

  // Set secure permissions
  try {
    chmodSync(dbPath, 0o600);
  } catch {
    // Ignore permission errors on Windows
  }

  globalSpendDb = db;
  return db;
};

/**
 * Closes the global spend database
 */
export const closeSpendDb = (): void => {
  if (globalSpendDb) {
    globalSpendDb.close();
    globalSpendDb = null;
  }
};

// ============================================================================
// Schema
// ============================================================================

const initSpendSchema = (db: Database.Database): void => {
  db.exec(`
    CREATE TABLE IF NOT EXISTS schema_version (
      version INTEGER PRIMARY KEY,
      applied_at INTEGER NOT NULL
    );

    CREATE TABLE IF NOT EXISTS spend_log (
      id INTEGER PRIMARY KEY AUTOINCREMENT,
      cost_usd REAL NOT NULL,
      repo_id TEXT,
      created_at INTEGER NOT NULL
    );

    CREATE INDEX IF NOT EXISTS idx_spend_log_created_at ON spend_log(created_at);
  `);

  // Check/update schema version
  const versionRow = db
    .prepare("SELECT COALESCE(MAX(version), 0) as version FROM schema_version")
    .get() as { version: number };

  if (versionRow.version < 1) {
    db.prepare(
      "INSERT INTO schema_version (version, applied_at) VALUES (?, ?)"
    ).run(1, Math.floor(Date.now() / 1000));
  }
};

// ============================================================================
// Spend Operations
// ============================================================================

/**
 * Records a spend amount to the global spend log.
 * repoId is optional and used for auditing purposes.
 */
export const recordSpend = (costUsd: number, repoId?: string): void => {
  if (costUsd <= 0) {
    return;
  }

  const db = getSpendDb();
  const stmt = db.prepare(
    "INSERT INTO spend_log (cost_usd, repo_id, created_at) VALUES (?, ?, ?)"
  );
  stmt.run(costUsd, repoId ?? null, Math.floor(Date.now() / 1000));
};

/**
 * Returns the total spend for the given month (format: "YYYY-MM").
 * If month is empty, uses the current month.
 */
export const getMonthlySpend = (month?: string): number => {
  const db = getSpendDb();

  // Parse or default to current month
  const targetMonth = month || formatMonth(new Date());
  const [startTs, endTs] = getMonthBounds(targetMonth);

  const result = db
    .prepare(
      "SELECT COALESCE(SUM(cost_usd), 0) as total FROM spend_log WHERE created_at >= ? AND created_at < ?"
    )
    .get(startTs, endTs) as { total: number };

  return result.total;
};

// ============================================================================
// Helper Functions
// ============================================================================

/**
 * Formats a date as "YYYY-MM"
 */
const formatMonth = (date: Date): string => {
  const year = date.getFullYear();
  const month = String(date.getMonth() + 1).padStart(2, "0");
  return `${year}-${month}`;
};

/**
 * Returns start and end timestamps for a month (format: "YYYY-MM")
 */
const getMonthBounds = (month: string): [number, number] => {
  const [yearStr, monthStr] = month.split("-");
  const year = Number.parseInt(yearStr ?? "", 10);
  const monthNum = Number.parseInt(monthStr ?? "", 10);

  if (Number.isNaN(year) || Number.isNaN(monthNum)) {
    throw new Error(`Invalid month format: ${month}. Expected YYYY-MM`);
  }

  const start = new Date(year, monthNum - 1, 1, 0, 0, 0, 0);
  const end = new Date(year, monthNum, 1, 0, 0, 0, 0); // First of next month

  return [Math.floor(start.getTime() / 1000), Math.floor(end.getTime() / 1000)];
};
